### PR TITLE
feat(fabric all in one): Runs a Fabric Network in one docker container

### LINKED ISSUE #132: Fabric 1.4.x All In One Docker Image
### Description
As a developer I want to have Cactus ship with an all in one Fabric 1.4.x image so that I can write (relatively) lightweight and fully automated test cases where a ledger is brought up from scratch, used for testing and then stopped and cleaned up. 

These files are examples of the same thing already done for Quorum and Besu in action (e.g. there are passing tests with contract deployment for them)

Ledger class:
`packages/bif-test-tooling/src/main/typescript/quorum/quorum-test-ledger.ts`

Test case using the Ledger class for test tooling:
`packages/bif-plugin-ledger-connector-quorum/src/test/typescript/integration/plugin-ledger-connector-quorum/deploy-contract/deploy-contract-from-json.ts`

The same classes exist for Besu.

An example of the all in one image for Besu can be seen here:
https://github.com/petermetz/besu-docker-all-in-one
Notice how it uses docker build stages to combine together orion and besu's base images.

One thing that is not yet done in the besu all in one docker image (but sorely needed) is the possibility to inject a custom genesis JSON object through code directly from the tests, meaning that a test could generate it's own list of test account addresses with specific balances to create any kind of scenario.

### Acceptance Criteria
1. `docker pull fabric-all-in-one` (or similar) works out of the box
1.1 Avoid using docker volumes and host ports by default to make sure container instantiation is re-entrant e.g. I can blindly start multiple containers with the same image with default options and it doesn't start crashing because of volume/port conflicts.
2. Have a class `FabricV1TestLedger` in the `test-tooling` package that is responsible for managing the all in one docker image (can start, stop, destroy it and allows for customization) similarly as the class linked above (`QuorumTestLedger`)
2. Customizable initial state through parameters of the constructor/start method of the `FabricV1TestLedger` class.
3. A test case in the `*test-tooling` package that verifies the image and the `FabricV1TestLedger` in action for deploying a hello world sort of contract. Very similar to this here: `packages/bif-plugin-ledger-connector-quorum/src/test/typescript/integration/plugin-ledger-connector-quorum/deploy-contract/deploy-contract-from-json.ts`

> FYI: I'm working on REST API endpoints that expose contract compilation as a feature. Starting with solidity but would be nice to have go as well for chaincode.

cc: @sownak @jonathan-m-hamilton

### PR DESCRIPTION
**Pull request by sownak**
Fixes #132 

Changes:
1. docker pull hyperledger/cactus-fabric-all-in-one should work after building and pushing.
2. No docker volume used.
3. A class FabricV1TestLedger in the cactus-test-tooling package added that is responsible for managing the all in one docker image (can start, stop, destroy it and allows for customization).

This is a draft PR following are pending:
1. Deploying a hello world sort of contract as Fabric contracts/chaincode are needed to be compiled on the running peer Container.

@petermetz

### CODE REVIEW

**Comment by RafaelAPB** [comment]
This looks amazing, @sownak! Do you have any ideas on how to adapt this connector to Fabric v2.0?

**Comment by sownak** [comment]
@RafaelAPB Just change the Fabric version in the Dockerfile and rebuild the image. Should work.

**Comment by sfuji822** [comment]
Thanks for your contribution. 
However, I prefer to having docker-compose style solution as same as other all-in-one containers like Bese or Corda instead of nodeJS application to be invoked.
I assume setting up channels and deploying chaincodes can be executed by admin user later, and it is more useful since chaincode is merely application in Cactus's context. 
I also wanted to remove files that are copyrighted by someone else, and you should instruct how to retrieve them by a user.
One more thing, used v1.4.4 is pretty old. v1.4.7 or v.1.4 (alias) should be chosen for safe.

**Comment by jonathan-m-hamilton** [comment]
July 20th Maintainers Meeting
* Discussed need to have both docker-compose & all-in-one solutions; not mutually exclusive so this PR can continue
* Need to validate why IBM copyright files are included rather than pulling at build / run-time
* Ideally should use latest Fabric version (unless Sownak confirms reason for certain version)

**Comment by sownak** [comment]
Comments noted. Will progress on them. Going for Fabric 1.4.7 because I believe we will have to provide separate code for 2.x.x has there are a few differences between 1.4.x and 2.x.x

File: packages/cactus-test-tooling/src/main/typescript/fabric/fabric-test-ledger-v1.ts [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+  containerImageVersion?: string;
+  containerImageName?: string;
+  restApiHttpPort?: number;
+}
+
+/*
+ * Provides default options for Fabric container
+ */
+export const FABRIC_TEST_LEDGER_DEFAULT_OPTIONS = Object.freeze({
+  containerImageVersion: "1.4.4",
```
[the remark this answers is not in the document]
Comment (RafaelAPB) [reply packages/cactus-test-tooling/src/main/typescript/fabric/fabric-test-ledger-v1.ts:20]: @sfuji822 I belive the latest is 1.4.8

[the remark this answers is not in the document]
Comment (sfuji822) [reply packages/cactus-test-tooling/src/main/typescript/fabric/fabric-test-ledger-v1.ts:20]: Then can you update with 1.4.8 :-) ?

[the remark this answers is not in the document]
Comment (sownak) [reply packages/cactus-test-tooling/src/main/typescript/fabric/fabric-test-ledger-v1.ts]: Updated to 1.4.8

**Comment by sownak** [comment]
> @sownak Regarding the code under `tools/docker/fabric-all-in-one/typescript/` (e.g. the NodeJS web application) => The functionality provided by this should be in one of the cactus packages.
> The (currently non-existent) ledger connector plugin package for `fabric` is the most fitting from the top of my head.
> Reasoning: If the ledger connector plugin depends on some random/custom made NodeJS app being deployed on the Fabric nodes that we intend to connect to through it, then that adds operational complexity and an actual migration step if you want to deploy Cactus to interact with an existing Fabric installation where you did not previously have this NodeJS app, the ports open for it, etc. In some places just getting the firewall rules changed can be something that takes weeks.
> The bet is that by relying only on SSH connections being available to the machines running the Fabric nodes we increase the chance of that dependency being satisfied by default to 99% from 0.1% with the NodeJS app.

@petermetz All changes related to NodeJs app is reverted. We should be creating a new Issue for that.

This PR contains Fabric 1.4.8 all-in-one docker image with single orderer, single peer and 1 channel created.

**Comment by petermetz** [comment]
> Comments noted. Will progress on them. Going for Fabric 1.4.7 because I believe we will have to provide separate code for 2.x.x has there are a few differences between 1.4.x and 2.x.x

Agreed, I heard the differences are pretty small, but unless there's full compatibility, we will need a separate one for 2.x (not a huge problem, it is a major release afterall)

**Comment by petermetz** [comment]
@sownak I published the image on docker hub as we discussed in today's call: https://hub.docker.com/repository/docker/hyperledger/cactus-fabric-all-in-one/tags?page=1

**Comment by sownak** [comment]
> @sownak LGTM, but please rebase to the master branch and squash everything into a single commit. PR name is a good start for the commit message but please avoid having space characters in the subject (stuff in between the parens) so something like this would be great (just added hyphens instead of spances)
> `feat(fabric-all-in-one): Runs a Fabric Network in one docker container`
> 
> If you need help with the git commit log surgery just ping me, happy to help.

Done.

File: packages/cactus-test-tooling/src/main/typescript/fabric/fabric-test-ledger-v1.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (petermetz) [reply packages/cactus-test-tooling/src/main/typescript/fabric/fabric-test-ledger-v1.ts]: Re-requesting review now that it's been updated

**Comment by sfuji822** [comment]
LGTM

**Pull request merged**